### ISSUE TITLE
Native Vim/Emacs controls

### ISSUE THREAD

**Issue by xbc5** [LadybirdBrowser/ladybird#981]
Please consider native (modal) Vim and Emacs controls for the _entire_ interface.

The current generation of Vim web-extension support (in other browser) is frustrating because one needs to wait for page loads to control web-page elements; and they cannot interface with the core browser.

Web extensions like [Tridactyl](https://github.com/tridactyl/tridactyl) offer a battle tested way to use Vim on the web -- it could serve as influence.

This is probably a low priority feature, but consideration early is appreciated.

Thanks!

**Comment by ActivexDiamond** [comment]
Would love to see this in LadyBird.

**Comment by KennethHoff** [comment]
Everytime I hear of a new browser, my first instinct is to check if they have native Vim support. While Vimium, Surfing Keys, Tridactyl, and similar are good, they often have severe limitations (doesn't work everywhere; not in vendor-specific menus, like about:*, only works after page load, etc..).

It would be lovely if Ladybird supported fully customizable keybindings, and in particular a Vim emulation mode.

**Comment by brtholomy** [comment]
keybinding configuration via a simple and separate file we can safely check in to a public repo (no mixing with login secrets/cookies etc):

```
~/.config/ladybird/keybindings.toml
```

would be ideal.

**Comment by lumosminima** [comment]
I started using vim/nvim a month ago and really the biggest benefit for me as a developer are its precise surgical motions, I use it inside vscode with flash.nvim plugin which allows cursor jump using vimium style labels and vscode-multi-cursor.nvim.  Something like this on ladybird would immediately make me convert to it since currently I have to resort to mouse since those vim style extensions are only good for link jumping ime and their selection and yanking is not as satisfying (search and then visual mode to yank, no visible cursor to adjust selection, no modes in text fields like this etc)

**Comment by xbc5** [comment]
> I started using vim/nvim a month ago and really the biggest benefit for me as a developer are its precise surgical motions, I use it inside vscode with flash.nvim plugin which allows cursor jump using vimium style labels and vscode-multi-cursor.nvim. Something like this on ladybird would immediately make me convert to it since currently I have to resort to mouse since those vim style extensions are only good for link jumping ime and their selection and yanking is not as satisfying (search and then visual mode to yank, no visible cursor to adjust selection, no modes in text fields like this etc)

The worst element about modern browsers is the laborious interface. I'd really love to see a new browser that has a different angle. There's so much potential there (going in a new direction) because Firefox, Chrome, and Safari are all the same, and have entirely stagnated with innovation.

**Comment by AtkinsSJ** [comment]
I just realised there's been no actual response from a maintainer yet, so:

Vim bindings goes in the category of features that we wouldn't object to someone adding (assuming it's not hugely invasive) but the core team is unlikely to work on, at least for the time being. Our focus is on trying to get the browser in a good enough state to meet the roadmap goals of a usable browser in 2026, and stable release a little later, and that's already a lot of work. 😅

This isn't meant to shoot the idea down, just set expectations. You're very welcome to have a go at it yourselves! I think you make a fair point about an extension not working too well. (Not that we have extension support yet either!)

**Comment by brtholomy** [comment]
We know building a usable browser is an enormous undertaking, and we're excited for it.

But consider your market realistically: what's your selling point? Not being another blink/gecko clone? Who cares about that except devs? And what do they want? Power features, customizability. Firefox has [alienated its core userbase](https://www.reddit.com/r/firefox/comments/1fu19hs/comment/lq1jv1p/) past redemption, so your timing is excellent.

We want this to happen and care, but we will drool over it if you give us poweruser features. In fact we'll put up with bad performance on YT/Netflix just for this. Consider the marketing hyprland and NixOS gets on r/unixporn : that's what you want, and nothing else will propel ladybird past the Chrome inertia. The larger userbase will follow the powerusers. My 2cts.

**Comment by xbc5** [comment]
> But consider your market realistically: what's your selling point? Not being another blink/gecko clone? Who cares about that except devs? And what do they want? Power features, customizability. 

Oh I agree, your core hardcore, early adopters are the people who market your product for you, for free: because they believe in it. Another generic browser the world does not need.

> Firefox has [alienated its core userbase](https://www.reddit.com/r/firefox/comments/1fu19hs/comment/lq1jv1p/) past redemption

Firefox died (for me) when it abandoned its own addon format (which it pioneered), in favour of web-extensions: I stayed with Firefox just for that feature (despite terrible performance); I went to Chrome when that disappeared. 

> Consider the marketing hyprland and NixOS gets on r/unixporn

Nobody markets towards the hardcore users, but there's a small but dedicated user base crying out for power features.

**Comment by brtholomy** [comment]
> It's a spark that may start a fire.

And the story goes deeper. We want this from at least 2 angles:

1. The internet is making us stupider. Like many linux users, I both love and hate computers. I stick to the commandline as much as possible, and despise the laborious browser interface. A browser that actually enabled the kind of athletic unconscious competence of typing everywhere in the interface, would change a great deal of the way information is currently ingested, especially relevant to the degradation of the attention span of children.

2. Anti-trust and preservation of free speech. The browser that pleases this type - and we are probably a very powerful 2-3% - will have a fighting chance at preventing or delaying the eventual internet lockdown. But it must possess a rendering engine powerful enough to keep the pace with Chromium, or it's just a toy. Ladybird looks to me to be the first project with that kind of horsepower.

**Comment by ADKaster** [comment]
The issue tracker is not a place to discuss broad patterns in the web browser ecosystem. Please focus this issue on a possible implementation of alternative keyboard bindings to the UI. As Sam mentioned earlier, this feature is not on the roadmap for the Alpha release in 2026, but a PR adding such a feature is welcome.

**Comment by bpsomu** [comment]
Would love it if Ladybird includes Vim-like keybindings in the long term. Have just started using Zed editor with Vim mode and man it's awesome.

**Comment by Glog78** [comment]
I love qutebrowser for the way it is configurable. Like lets be honest -> something like  https://wiki.archlinux.org/title/Qutebrowser#Video_playback is just amazing. What i am missing and most likely will never be included is ... hint <scrollable_area> select_scrollable_area {hint-area} , just because the underlying engine won't expose it and css doesn't have a good way to select this. So if you decide to have such an input system this are two of the things i would love to see come out of this.

**Comment by bpsomu** [comment]
> You're very welcome to have a go at it yourselves!

Recently Ladybird project decided to not take contributions from outside because of AI slop. While I understand the decision, I am quite saddened by it. Hate the negative effect AI is having on FOSS.
 
Where does this new decision leave us with features like this that are unlikely to be implemented by the main project team?

**Comment by gmta** [comment]
> > You're very welcome to have a go at it yourselves!
> 
> Recently Ladybird project decided to not take contributions from outside because of AI slop.

That's not entirely correct, please read the announcement carefully.

> While I understand the decision, I am quite saddened by it. Hate the negative effect AI is having on FOSS.
> 
> Where does this new decision leave us with features like this that are unlikely to be implemented by the main project team?

In the future, browser extensions will need to be supported. Relatively esoteric features like vim/emacs controls support belong with those imo.

**Comment by gmta** [comment]
Closing this as not planned.

**Comment by darukutsu** [comment]
> In the future, browser extensions will need to be supported. Relatively esoteric features like vim/emacs controls support belong with those imo.

Well depends if you plan to implement extension security like firefox did, because then injection like vimium or tridactyl extensions do will be limited unless proper API is implemented, it would not work on all pages and user has to wait before page is loaded to use those binding. Other browsers like qutebrowser also expose env variables containing URL, selected text, tab index, webpage content etc. that you can then manipulate with in script and create custom chained commands.

And since none of ladybird developers seem to understand benefit of having this kind of ability this was great task for community to implement. It is saddening that this project choose route of truly not "open source" but more "source available" (with exception that anyone can fork it and continue community route).

For me this project adds at this current state 0 benefit over qutebrowser, firefox or even chromium and likely will be just another mouse driven browser.

**Comment by gmta** [comment]
Yeah, we're not rehashing discussions that could be had over at our Discord. Locking the convo.
